src/main/java/org/takes/rq/ChunkedInputStream.java

```diff
@@ -0,0 +1,253 @@
+ *
+ * @author Maksimenko Vladimir (xupypr@xupypr.com)
+ * @version $Id$
+ * @since 0.31.2
+ * @checkstyle ExecutableStatementCountCheck (500 lines)
+ * @checkstyle CyclomaticComplexityCheck (500 lines)
+ * @checkstyle LineLengthCheck (1 lines)
+ * @link <a href="https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.6.1">Chunked Transfer Coding</a>
+ */
+@SuppressWarnings
+    (
+        {
+            "PMD.CyclomaticComplexity",
+            "PMD.StdCyclomaticComplexity"
+        }
+    )
+final class ChunkedInputStream extends InputStream {
+
+    /**
+     * The inputstream that we're wrapping.
+     */
+    private final transient InputStream origin;
+
+    /**
+     * The chunk size.
+     */
+    private transient int chunk;
+
+    /**
+     * The current position within the current chunk.
+     */
+    private transient int pos;
+
+    /**
+     * True if we'are at the beginning of stream.
+     */
+    private transient boolean bof;
+
+    /**
+     * True if we've reached the end of stream.
+     */
+    private transient boolean eof;
+
+    /**
+     * Ctor.
+     * @param stream The raw input stream
+     * @throws IOException If an IO error occurs
+     */
+    ChunkedInputStream(final InputStream stream) throws IOException {
+        super();
+        this.bof = true;
+        this.origin = stream;
+    }
+
+    @Override
+    public int read() throws IOException {
+        if (!this.eof && this.pos >= this.chunk) {
+            this.nextChunk();
+        }
+        final int result;
+        if (this.eof) {
+            result = -1;
+        } else {
+            ++this.pos;
+            result = this.origin.read();
+        }
+        return result;
+    }
+
+    @Override
+    public int read(final byte[] buf, final int off, final int len)
+        throws IOException {
+        if (!this.eof && this.pos >= this.chunk) {
+            this.nextChunk();
+        }
+        final int result;
+        if (this.eof) {
+            result = -1;
+        } else {
+            final int shift = Math.min(len, this.chunk - this.pos);
+            final int count = this.origin.read(buf, off, shift);
+            this.pos += count;
+            if (shift == len) {
+                result = len;
+            } else {
+                result = shift + this.read(buf, off + shift, len - shift);
+            }
+        }
+        return result;
+    }
+
+    @Override
+    public int read(final byte[] buf) throws IOException {
+        return this.read(buf, 0, buf.length);
+    }
+
+    /**
+     * Read the CRLF terminator.
+     * @throws IOException If an IO error occurs.
+     */
+    private void readCRLF() throws IOException {
+        final int crsymbol = this.origin.read();
+        final int lfsymbol = this.origin.read();
+        if (crsymbol != '\r' || lfsymbol != '\n') {
+            throw new IOException(
+                "CRLF expected at end of chunk: " + crsymbol + "/" + lfsymbol
```

Comment:
@xupyprmv string concatenation is a bad idea, see http://www.yegor256.com/2014/06/19/avoid-string-concatenation.html